### ISSUE TITLE
Problem when installing via pip on Windows or compiling from source

### ISSUE THREAD

**Issue by blais** [beancount/beancount#311]
**[Original report](https://bitbucket.org/blais/beancount/issue/311) by Victor Araújo (Bitbucket: [victoraraujo01](https://bitbucket.org/victoraraujo01), GitHub: [victoraraujo01](https://github.com/victoraraujo01)).**

----------------------------------------

Python version: 3.7.0
Pip version: 10.0.1

Steps to reproduce:

1. Install python from .msi
2. Install Visual Studio Code Build Tools
3. Try to install beancount either via pip or by compiling the source code

Errors obtained:

`beancount/parser/parser.c(290): warning C4003: not enough actual parameters for macro 'STRINGIFY'`

Intepretation:

Looking at the whole installation log I found two other errors related to version control. Since my system is in portuguese, I'll try to translate it:

    aborted: repository not found in 'C:\Users\u4dd\AppData\Local\Temp\pip-install-o355xqsy\beancount' (.hg not found)!
    fatal: not a git repository (or any of the parent directories): .git

When compiling from source, the error is a little different:

    hg: input analysis error: date expects date information
    fatal: not a git repository (or any of the parent directories): .git

So, from what I could tell, because the calls to `get_git_changeset` and  `get_hg_changeset` fail, the `vc_changeset` variable on the `setup.py` file ends up being set to an empty string `''`. Because it is empty, the `setup()` call does not set the `VC_CHANGESET` macro and it causes an error on the compilation of the `parser.c` file.

The calls to `get_git_changeset` and  `get_hg_changeset` fail on the installation via pip probably because the package does not contain a `.git` or `.hg` directory. However, when trying to compile from source, I've noticed that the Mercurial command it tries to run appears to be incorrect, since `"%s"` is not a valid option. Maybe it's supposed to be in caps, like `%S`?

**Comment by blais** [comment]
**Original comment by Zach Smith (Bitbucket: [z__j](https://bitbucket.org/z__j)).**

----------------------------------------

I believe the issue is with a quirk of Visual Studio's preprocessor.  Somehow stringizing the null parameter gets reduced to empty text and not an empty string.  For example, this code (invoked with preprocessing only):
```
#!c
#define XSTRINGIFY(x) STRINGIFY(x)
#define STRINGIFY(x) #x

result = XSTRINGIFY(VC_CHANGESET);
```
 produces, with GCC:

```
$ gcc -E test.c
# 1 "test.c"
# 1 "<built-in>"
# 1 "<command-line>"
# 1 "/usr/include/stdc-predef.h" 1 3 4
# 1 "<command-line>" 2
# 1 "test.c"



result = "";
```
while with Visual Studio:

```
$ ./x64/cl.exe -E -DVC_CHANGESET=  test.c
Microsoft (R) C/C++ Optimizing Compiler Version 19.13.26129 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

test.c
test.c(4): warning C4003: not enough actual parameters for macro 'STRINGIFY'
#line 1 "test.c"



result = ;

```

A possible fix is to use the compiler's string concatenation to hide this detail, probably just for Visual Studio to prevent the 'mistake' from getting deleted later.
For instance, lines 287ff. could be updated to read:
```
#!c

#ifdef _MSC_VER
    // Extra Null string is a Visual Studio workaround, do not remove
    static const char* vc_changeset_str = "" XSTRINGIFY(VC_CHANGESET); 
#else
```
Which is actually pretty ugly.   Visual Studio will still emit warning C4003, but the final result will still be valid C.

But it does work and keeps the intended behavior in invocations with and without VC_CHANGESET:

```
$ ./x64/cl.exe -E -DVC_CHANGESET=  fixed.c
Microsoft (R) C/C++ Optimizing Compiler Version 19.13.26129 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

fixed.c
fixed.c(4): warning C4003: not enough actual parameters for macro 'STRINGIFY'
#line 1 "fixed.c"



result = "" ;

$ ./x64/cl.exe -E -DVC_CHANGESET=hg.deadbeef0123456789  fixed.c
Microsoft (R) C/C++ Optimizing Compiler Version 19.13.26129 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

fixed.c
#line 1 "fixed.c"



result = "" "hg.deadbeef0123456789";

```

**Comment by blais** [comment]
**Original comment by Zach Smith (Bitbucket: [z__j](https://bitbucket.org/z__j)).**

----------------------------------------

> However, when trying to compile from source, I've noticed that the Mercurial command it tries to run appears to be incorrect, since "%s" is not a valid option. Maybe it's supposed to be in caps, like %S?

This is because `%s` isn't an officially-supported Python format string, and whatever implementation of strfmtime we've got here on Windows doesn't have it.  See [this SO answer](https://stackoverflow.com/a/11743262).

Since `hgdate` already gives the UTC timestamp, changing `setup.py` to request this and just trimming off the timezone offset in Python, post-strip(), works (at least on my system):
```
#!python
output = subprocess.check_output(
    ['hg', 'parent', '--template', '{node} {date|hgdate}'], shell=False)
vc_changeset, vc_timestamp = output.decode('utf-8').split()[:2]
```

Cross-ref #304 where this formatting got introduced

**Comment by blais** [comment]
**Original comment by Martin Blais (Bitbucket: [blais](https://bitbucket.org/blais), GitHub: [blais](https://github.com/blais)).**

----------------------------------------

Fix #311: Bypass MSVC quirk; missing strfmtime formats

For `#define STRINGIFY(s) #s` Visual Studio's preprocessor does
    `STRINGIFY()` -> <NOTHING>
and not the desired
    `STRINGIFY()` -> `""`

As a workaround, pre-insert an empty string if we detect we're in
MSVC-land, and let the compiled remove it.  If the stringified
argument is empty this prevents a syntax error.

The other problem: Windows's implementation of `strfmtime` apparently
lacks the `%s` format code used to get a timestamp.  Switch to
trimming the timezone part in Python, which should be cross-platform.
